### DISCUSSION TITLE
How to manage the query key correctly

### DISCUSSION THREAD

**Discussion by tianleiyiji123** [TanStack/query#2067] [Q&A]
How to effectively and reasonably manage the key field in each query? When creating a query, the query may change according to the parameters, and then when you want to use the query again on another page, the key cannot be clearly known

**Comment by TkDodo** [comment]
I usually manage the keys in global client state, or in router state, like:

`myUrl/todos/1`

if I switch to a different todo, I update the url:

`myUrl/todos/5`

and all subscribers will re-render and re-fetch (if you use something like react-router).

so the url is a good global state manager for these kinds of things

**Reply by tianleiyiji123** [reply]
Yes, I usually define the key of the query in this way, but there is a problem, that is, when a query is defined on a page, the key of the query is changed according to the request parameters, and then there is a mutation operation on another page to complete To update the list on the previous page after the operation, I need to clear the cache and keep the list up to date. At this time, I don’t know which key is the previous page.

**Reply by tianleiyiji123** [reply]
![image](https://user-images.githubusercontent.com/13405335/113373974-98a50700-939e-11eb-9b10-3b22c02d2484.png)
I will wrap the query in a hook. I want to call this method directly to clear the cache, but because there are parameters, I don’t know which key’s cache needs to be cleared on other pages.

**Reply by TkDodo** [reply]
So you perform a mutation, and then redirect to a different page before the mutation is finished?

**Reply by tianleiyiji123** [reply]
No, for example, from a communication list page, each list is a communication task. Click the button to communicate, and it will jump to the communication details page. When the communication is completed, I want to trigger invalidateQueries to update the list, but at this time I It is impossible to know the key of the previous list query on the details page.

**Reply by TkDodo** [reply]
it sounds like list and detail are clearly related, so for example:

communication-list-key: `['communications']`
communication-detail-key (with id: 15): `['communications', 15]`

so it's `invaliateQueries(['communications'])` in this case. I'm clearly not getting the full picture - maybe some example runnable codesandbox code would help

**Reply by tianleiyiji123** [reply]
Yes, you can succeed in this way. But when it comes to key maintenance in a relatively large project, I think we need to find a good maintenance method. As you said, the key is maintained through routing, but in this way, when you need to operate the query in other components or other pages, you will use the key repeatedly to confirm a query, which is not convenient at this time.
